File: web/src/utils/moneyParity.test.ts
import { readFileSync } from 'node:fs';
import { resolve } from 'node:path';
import { describe, expect, it } from 'vitest';
import { formatAmount, formatAmountForInput, parseAmount } from '@/utils/currency';
import { formatCsvAmount } from '@/utils/analytics';

const fixture = JSON.parse(
  readFileSync(resolve(process.cwd(), '../testdata/money-parity.json'), 'utf8'),
) as {
  display: Array<{
    amount: number;
    currency: string;
    language: 'en' | 'de';
    showSymbol: boolean;
    formatted: string;
  }>;
  parse: Array<{ input: string; currency: string; value: number }>;
  csv: Array<{ amount: number; cell: string }>;
  input: Array<{ amount: number; currency: string; text: string }>;
};

describe('money-parity.json (shared with Android MoneyParityTest)', () => {
  it('display matches app-language formatting', () => {
    for (const row of fixture.display) {
      expect(formatAmount(row.amount, row.currency, row.showSymbol, row.language)).toBe(row.formatted);
    }
  });

  /**
   * Edit-form prefill. Android rendered this with DecimalFormat("0.##"), which drops
   * trailing zeros, so a stored 12.50 prefilled "12,5" there and "12,50" here — the same
   * transaction shown differently depending on which client opened it.
   */
  it('edit-form prefill matches Android formatAmountForInput', () => {
    for (const row of fixture.input) {
      expect(formatAmountForInput(row.amount, row.currency)).toBe(row.text);
    }
  });

  it('parse matches Android CurrencyUtils vectors', () => {
    for (const row of fixture.parse) {
      expect(parseAmount(row.input, row.currency)).toBe(row.value);
    }
  });

  it('csv cells are two-decimal US amounts', () => {
    for (const row of fixture.csv) {
      expect(formatCsvAmount(row.amount)).toBe(row.cell);
    }
  });
});

File: app/src/test/java/com/aus/ausgegeben/util/MoneyParityTest.kt
package com.aus.ausgegeben.util

import org.json.JSONObject
import org.junit.Assert.assertEquals
import org.junit.Test

/**
 * Loads testdata/money-parity.json (same file as web/src/utils/moneyParity.test.ts).
 */
class MoneyParityTest {

    private val fixture: JSONObject = JSONObject(
        checkNotNull(javaClass.classLoader!!.getResourceAsStream("money-parity.json"))
            .bufferedReader()
            .use { it.readText() },
    )

    @Test
    fun display_matchesWebLanguageRule() {
        val rows = fixture.getJSONArray("display")
        for (i in 0 until rows.length()) {
            val row = rows.getJSONObject(i)
            assertEquals(
                row.getString("formatted"),
                CurrencyUtils.formatAmount(
                    row.getDouble("amount"),
                    row.getString("currency"),
                    showSymbol = row.getBoolean("showSymbol"),
                    language = row.getString("language"),
                ),
            )
        }
    }

    @Test
    fun parse_matchesWebVectors() {
        val rows = fixture.getJSONArray("parse")
        for (i in 0 until rows.length()) {
            val row = rows.getJSONObject(i)
            assertEquals(
                row.getDouble("value"),
                CurrencyUtils.parseAmount(row.getString("input"), row.getString("currency"))!!,
                0.001,
            )
        }
    }

    @Test
    fun csv_matchesWebTwoDecimalCells() {
        val rows = fixture.getJSONArray("csv")
        for (i in 0 until rows.length()) {
            val row = rows.getJSONObject(i)
            assertEquals(row.getString("cell"), ExportUtils.formatAmountCell(row.getDouble("amount")))
        }
    }
    /**
     * Edit-form prefill. This used DecimalFormat("0.##"), which drops trailing zeros, so
     * a stored 12.50 prefilled "12,5" here and "12,50" on web -- the same transaction
     * shown differently depending on which client opened it. Both re-parsed correctly, so
     * nothing was miscalculated; it was a visible parity gap.
     */
    @Test
    fun inputPrefill_matchesWebFormatAmountForInput() {
        val rows = fixture.getJSONArray("input")
        for (i in 0 until rows.length()) {
            val row = rows.getJSONObject(i)
            assertEquals(
                row.getString("text"),
                CurrencyUtils.formatAmountForInput(row.getDouble("amount"), row.getString("currency")),
            )
        }
    }
}
